only query for db-metadata once per sync operation + scan each table for its PKs

escherize opened the pull request:
fixes #26054

# Part 1:

I've solved the excessive re-querying issue by querying for `db-metadata` up front, and passing the data into `sync-tables` and `metabase-metadata`.

```mermaid

graph TD
POST_:id/sync_schema --> api/write-check
POST_:id/sync_schema --> future
future --> sync-metadata/sync-db-metadata!
sync-metadata/sync-db-metadata! --> sync-operation
sync-operation --> timezone
sync-operation --> tables
sync-operation --> fields
sync-operation --> fks
sync-operation --> metabase-metadata

future --> analyze/analyze-db!

analyze/analyze-db! --> sync-operation2[sync-operation]
sync-operation2[sync-operation] --> fingerprint-fields
sync-operation2[sync-operation] --> classify-fields
sync-operation2[sync-operation] --> classify-tables


style tables color:#f22
style metabase-metadata color:#f22
```

`sync-tables` and `metabase-metadata` both call `describe-tables`, but we should only call it once for the sync operation.

---

# Part 2

The Snowflake JDBC driver ignores the table in `metadata.getPrimaryKeys(catalog, schema, table)`, unless `catalog` and `schema` are both passed.

Passing all 3 forces the Snowflake JDBC driver to issue the proper query. Here's a screenshot of the snowflake activity log:

<img width="500" alt="Screen Shot 2022-12-01 at 3 33 56 PM" src="https://user-images.githubusercontent.com/343288/205173363-a31430a7-d9cd-4d90-88f3-ac9acfba4ea9.png">

(For reference, this used to issue the query: `show primary keys in account;`).

---

# Part 3

I couldn't reproduce this, and didn't attempt a fix.

escherize committed:
memoize snowflake describe-database w/ 5 min ttl

escherize commented:
Not sure if 5 minutes is right here, but this will resolve both: redundant ❄️ queries, and the race condition.

metamben commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -283,12 +284,12 @@
 
-(defmethod driver/describe-database :snowflake
-  [driver database]
+(defn describe-database* [database]
```
Might make sense to keep the `driver` parameter, so that drivers inheriting from snowflake can profit from it.

metamben reviewed (commented):
Especially if consistency (i.e., that the table descriptions are equal) is important, wouldn't be better to make `sync-tables` and `metabase-metadata` get the result of `describe-tables` as input parameter? This solution feels more like a work around than a real fix.

escherize committed:
pass in db-metadata where it is used,

- rather than querying for it in random functions

escherize committed:
Merge branch 'master' into bcm-snowflake-efficient-sync

escherize commented:
I'll reopen this tomorrow

escherize committed:
update callsite of sync-tables-and-database!

escherize committed:
Merge branch 'bcm-snowflake-efficient-sync' of github.com:metabase/metabase into bcm-snowflake-efficient-sync

escherize committed:
Merge branch 'master' into bcm-snowflake-efficient-sync

escherize committed:
let sync-metabase-metadata! look up db-metadata

- only when needed

escherize committed:
Merge branch 'bcm-snowflake-efficient-sync' of github.com:metabase/metabase into bcm-snowflake-efficient-sync

escherize committed:
pass catalog (db), and schema to .getPrimaryKeys

escherize committed:
Merge branch 'master' into bcm-snowflake-efficient-sync

escherize committed:
let describe-table* add in the nil table-name

escherize committed:
Merge branch 'bcm-snowflake-efficient-sync' of github.com:metabase/metabase into bcm-snowflake-efficient-sync

escherize committed:
cleanup

escherize committed:
use :snowflake instead of driver for inheritance

escherize committed:
concurrent-sync-test should call describe-database once

metamben commented on src/metabase/sync/sync_metadata.clj:
```diff
@@ -32,25 +33,26 @@
        total-fks updated-fks total-failed))
 
-(def ^:private sync-steps
+(defn ^:private make-sync-steps [db-metadata]
```
```suggestion
(defn- make-sync-steps [db-metadata]
```

metamben commented on test/metabase/sync/util_test.clj:
```diff
@@ -54,9 +54,8 @@
-        ;; Check the number of syncs that took place. Should be 2 (just the first)
-        (is (= 2
-               @calls-to-describe-database))))))
+        ;; Check the number of syncs that took place. Should be 1 (just the first)
```
This comment is no more helpful than the actual check following it and I don't quite understand "just the first". I would remove the comment.

metamben reviewed (approved):
LGTM. My only concern is that making the `.getPrimaryKeys` calls with dbname and schema might break other drivers. If the tests pass, it's good, IMO.

luizarakaki reviewed (changes requested):
@escherize It is now running the sync only once 🙌 , but I'm getting this exception when trying to sync FK
Using these [credentials](https://start.1password.com/open/i?a=AEO4BXOSHNFCJLK26XA7AD522E&v=fbhwdxcon6hj77du2c5cc7wvaa&i=kzp4a5bx4pnhv4z6pabmbn5awa&h=my.1password.com)

```
[backend] 2022-12-02 14:38:51,877 INFO sync.util :: STARTING: step 'sync-fields' for snowflake Database 5 'Snowflake Dev'
[backend] 2022-12-02 14:38:52,361 WARN sync.util :: Error syncing Fields for Table 'Table 92 'PUBLIC.TIMESTAMP_TEST''
[backend] clojure.lang.ArityException: Wrong number of args (2) passed to: metabase.driver.sql-jdbc.sync.describe-table/add-table-pks
```

escherize committed:
cleaning up

escherize committed:
call add-table-pks with 3 args

escherize committed:
call add-table-pks with 3 args

escherize committed:
try escaping entity names to appease Oracle Driver

- Oracle wants the table-name to be escaped
- escape is usually a no-op

escherize committed:
pass driver to escape entity

escherize committed:
Merge branch 'master' into bcm-snowflake-efficient-sync

luizarakaki reviewed (approved):
LGTM!

escherize commented:
Once I get the `Oracle` test working, I'll go ahead and merge this.

escherize committed:
pull get-table-pks into a multimethod,

- handle the discrepancy directly in Oracle driver

escherize committed:
remove inline def

escherize committed:
fix colliding consistent-namespace linting

- metabase.driver.sql-jdbc.common and
  metabase.driver.sql-jdbc.sync.common were
  mapped to sql-jdbc.common
- metabase.driver.sql-jdbc.sync.common is now
  mapped to sql-jdbc.sync.common

escherize committed:
fix reflection warnings

escherize committed:
apply fix for getting oracle pks

escherize committed:
nix an unused arg

escherize committed:
Merge branch 'master' into bcm-snowflake-efficient-sync

escherize committed:
Merge branch 'master' into bcm-snowflake-efficient-sync
